gdx/src/com/badlogic/gdx/utils/JsonSkimmer.java

```diff
@@ -0,0 +1,592 @@
+			69, 69, 3, 70, 70, 3, 8, 8, 71, 8, 3, 72, 72, 73, 72, 3, 3, 3, 0};
+	}
+
+	private static final byte _json_indicies[] = init__json_indicies_0();
+
+	private static byte[] init__json_trans_targs_0 () {
+		return new byte[] {35, 1, 3, 0, 4, 36, 36, 36, 36, 1, 6, 5, 13, 17, 22, 37, 7, 8, 9, 7, 8, 9, 7, 10, 20, 21, 11, 11, 11, 12,
+			17, 19, 37, 11, 12, 19, 14, 16, 15, 14, 12, 18, 17, 11, 9, 5, 24, 23, 27, 31, 34, 25, 38, 25, 25, 26, 31, 33, 38, 25, 26,
+			33, 28, 30, 29, 28, 26, 32, 31, 25, 23, 2, 36, 2};
+	}
+
+	private static final byte _json_trans_targs[] = init__json_trans_targs_0();
+
+	private static byte[] init__json_trans_actions_0 () {
+		return new byte[] {13, 0, 15, 0, 0, 7, 3, 11, 1, 11, 17, 0, 20, 0, 0, 5, 1, 1, 1, 0, 0, 0, 11, 13, 15, 0, 7, 3, 1, 1, 1, 1,
+			23, 0, 0, 0, 0, 0, 0, 11, 11, 0, 11, 11, 11, 11, 13, 0, 15, 0, 0, 7, 9, 3, 1, 1, 1, 1, 26, 0, 0, 0, 0, 0, 0, 11, 11, 0,
+			11, 11, 11, 1, 0, 0};
+	}
+
+	private static final byte _json_trans_actions[] = init__json_trans_actions_0();
+
+	private static byte[] init__json_eof_actions_0 () {
+		return new byte[] {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+			1, 0, 0, 0};
+	}
+
+	private static final byte _json_eof_actions[] = init__json_eof_actions_0();
+
+	static final int json_start = 1;
+	static final int json_first_final = 35;
+	static final int json_error = 0;
+
+	static final int json_en_object = 5;
+	static final int json_en_array = 23;
+	static final int json_en_main = 1;
+
+	// line 269 "JsonSkimmer.rl"
+
+	private boolean stop;
+
+	/** Causes parsing to stop after the current or next object, array, or value. */
+	public void stop () {
+		stop = true;
+	}
+
+	public boolean isStopped () {
+		return stop;
+	}
+
+	private String unescape (String value) {
+		int length = value.length();
+		StringBuilder buffer = new StringBuilder(length + 16);
+		for (int i = 0; i < length;) {
+			char c = value.charAt(i++);
+			if (c != '\\') {
+				buffer.append(c);
+				continue;
+			}
+			if (i == length) break;
+			c = value.charAt(i++);
+			if (c == 'u') {
+				buffer.append(Character.toChars(Integer.parseInt(value.substring(i, i + 4), 16)));
+				i += 4;
+				continue;
+			}
+			switch (c) {
+			case '"':
+			case '\\':
+			case '/':
+				break;
+			case 'b':
+				c = '\b';
+				break;
+			case 'f':
+				c = '\f';
+				break;
+			case 'n':
+				c = '\n';
+				break;
+			case 'r':
+				c = '\r';
+				break;
+			case 't':
+				c = '\t';
+				break;
+			default:
+				throw new SerializationException("Illegal escaped character: \\" + c);
+			}
+			buffer.append(c);
+		}
+		return buffer.toString();
+	}
+
+	protected void startObject (@Null String name) {
+	}
+
+	protected void startArray (@Null String name) {
+	}
+
+	protected void pop () {
+	}
+
+	protected void value (@Null String name, String value, boolean unquoted) {
+	}
```

Comment:
These seem to be intended to be implemented by subclasses as needed, but without any documentation, it's hard to know when these get called, what they do, or what they can be used to do. I'm guessing a name that is null indicates there was no name in that case.

Reply:
I've added javadocs. They are for event based parsing where you handle JSON constructs as they are encountered.

I also changed to `push` with a `boolean object`. I'm torn which I like better. The single `push` method is shorter when you only count the level for both objects and arrays, which is common.

I added a real world usage example. It's maybe a bit of a verbose/nasty way to parse JSON, but it's pretty fast (I say having never written benchmarks) and beats trying to parse with regex.

Reply:
Hooray docs! Hooray usage example! Definitely beats parsing JSON with regex.